Add warning message for deprecated "hashicorp/terraform" provider

kislerdm opened the pull request:
Closes #108 

## What changed

- Added the warning message for deprecated "hashicorp/terraform" provider

## Why do we need it

- Parity terraform and OpenTofu.

## Followups

- Consider the ports-adapters pattern to facilitate testability, i.e. define adapters to the clients using interfaces.

- Add more unit tests. Even if the logic seems simple, there is still a handful of conditions, we can do better when it comes to testing.

kislerdm commented:
@Yantrio @yaronya Hey folks! Could you please have a look? Thanks!

Yantrio commented:
How would you feel about moving this logic out to it's own package?

I think a small package with a function that accepts a namespace + type, and returns errors based on that would be amazing.

In the long run this package could be swapped out easily for something else.

Part of the reason that I raise this is that I feel like it doesn't really belong as a function dangling on `ListProvidersPathParams`

kislerdm commented:
@Yantrio Hey James! Thanks for a prompt review.

> How would you feel about moving this logic out to it's own package?

It's a very reasonable proposal indeed. I'll do the factoring according to it. 

Also, post-alpha, we shall how to govern warnings for any arbitrary provider. In other words, how can we make sure that we, as a governing body between the provider's user and the provider's maintainers, keep warnings and any additional provider's info consistently with intentions of the provider's maintainer. For example, shall we warn tofu users when the provider they rely on is deprecated a/k/a fetched from the archived github repo?

> Part of the reason that I raise this is that I feel like it doesn't really belong as a function dangling on ListProvidersPathParams

You are absolutely right that `ListProvidersPathParams` indicates the Rest interface, and as such it shall not be coupled with the logic coupled to the "`Provider`" object. 

WDYT about the module refactoring after the alpha release? I'd like to propose its simplification and clear boundary separation between every app layer. Frankly, our app feel over-complex despite the size of the problem it solves. I highlighted the proposal in the PR description as 

> Consider the ports-adapters pattern to facilitate testability, i.e. define adapters to the clients using interfaces.

The motivation:
- Maintainability;
- Testability;
- Interoperability through reuse of the packages in other modules, e.g. for CLI we could use in CI.

Yantrio commented:
I like this approach, especially passing the warnings through the context to simplify the function sigs everywhere. I'm happy to merge this.

When it comes to the post-alpha comments above and the TODOs in your code, I think we should raise some issues on this repo so we don't forget them for the long term solution, and each of those can be considered individually at a later date.

As for refactoring, I think it's a good idea and we should look into this, can you raise an issue with some of your comments around this please?

Yantrio commented:
@kislerdm Would you mind fixing the linting issues before we merge please? 🙏 Thanks

Yantrio commented on src/internal/warnings/warnings.go:
```diff
@@ -0,0 +1,49 @@
+// fmt.Println(warn)
+// >> [This provider is archived and no longer needed. The terraform_remote_state data source is built into the latest OpenTofu release.]
+func ProviderWarnings(providerNamespace, providerType string) []string {
+	switch providerNamespace {
```
linting is failing here, but i think the switch is the right way to go.

Would you mind adding a nolint comment here with some explaination around this?

kislerdm replied:
@Yantrio Thank you for your help! The proposal was implemented.

kislerdm commented:
@Yantrio 

> Would you mind fixing the linting issues before we merge please? 🙏 Thanks

Definitely! I also created the [issue "placeholder"](https://github.com/opentofu/registry/issues/112) to configure linting explicitly.

Zanda256 commented on src/internal/warnings/warnings.go:
```diff
@@ -0,0 +1,48 @@
+	return nil
+}
+
+var warningsContext = struct{}{} //nolint:gochecknoglobals // This is a commonly used pattern for context binding
```
Would it be better if this var was called `warningsContextKey`? I think that would be easier to understand.

kislerdm replied:
Hey! Thanks for suggestion, it does make sense indeed! I've implemented the change accordingly, however the key is called `contextKey` because it's obvious that it relates to warnings as it's internal to the `warnings` package. Please review the change and resolve the comment. Thanks!

Zanda256 reviewed (commented):
Minor change request

kislerdm commented:
@cube2222 hey Kuba, I support your point of view on explicit declaration, however context binding is a commonly used pattern to propagate information without modifying signatures. In this particular case, I was more inclined towards segregating the warnings handling logic without extending functions' signatures because there are many assumptions about warnings handling/reporting. In other words, I wanted to modify signatures of the least number of functions even though they are private.

All in all, if you'd find this a blocker, I can, of course, refactor. However, I'd rather suggest coming back to this decision post alpha release, because if I understood your comment [here](https://github.com/opentofu/registry/issues/113#issuecomment-1742865771), we will have to make more critical decisions about registry. WDYT?

kislerdm commented:
BTW the tf diagnostics is also singleton, but with a better defined concern. We could split warnings propagation into a dedicated diag context moving forward. For example, it'd enable us to report if the provider's repo was archived - find details here.

maunzCache commented on src/internal/warnings/warnings.go:
```diff
@@ -0,0 +1,48 @@
+
+import "context"
+
+// TODO: add warnings for any archived providers based on the github repo status
```
Not a big issue here but what about moving the TODOs into proper Github issues or resolving them before sending this into code nirvana :)?

kislerdm replied:
@maunzCache Hey! Thanks for your suggestion. I implemented it: as the issues were raised already, I reckon TODOs may indeed create a wrong impression of solution readiness :) Please resolve the conversation if the change is satisfactory. Thanks!

maunzCache replied:
I have not been digging the details. If the change is fine for you, it will be fine for me i guess. Also: We have git tracking the changes ;)

kislerdm commented:
Hey folks! I refactored the implementation for the sake of simplicity:

- Removed implicit warnings propagation in the context. I do agree with Kuba that decision for implicit warnings/diag propagation is premature.
- Refactored the initial logic to make the cache-aside strategy more clear for the code reader.
- Unified signatures of the "helper functions" which wrap the singleton `config`.

cc: @Yantrio @cube2222

cube2222 reviewed (approved):
This looks much simpler to me now!

I'll let @Yantrio make the final call though.

kislerdm commented:
Hey @Yantrio! Not at all! I'll fix linting error tomorrow.

kislerdm commented:
@Yantrio Hey James! The fix addressing the nil-pointer issue was implemented. Please have a look. Thanks!
